File: clients/node/fastkv/tests/test_integration.js
'use strict';

// Force stdout unbuffered when piped (needed for CI/scripts)
if (process.stdout && process.stdout._handle && typeof process.stdout._handle.setBlocking === 'function') {
  process.stdout._handle.setBlocking(true);
}

/**
 * FastKV Node.js client — integration tests.
 *
 * Run:  node test_integration.js [host] [port]
 * Defaults: host=127.0.0.1  port=8379
 */

const net = require('net');
const { FastKVClient } = require('../client');

// Helper: connect a client (replaces old `await new FastKVClient(...)`)
async function makeClient() {
  return FastKVClient.create({ host: HOST, port: PORT });
}

const HOST = process.argv[2] || '127.0.0.1';
const PORT = parseInt(process.argv[3] || '8379', 10);

let passed = 0;
let failed = 0;

function check(name, condition, msg) {
  if (condition) {
    passed++;
    console.log(`  PASS  ${name}`);
  } else {
    failed++;
    console.log(`  FAIL  ${name} — ${msg || ''}`);
  }
}

// ── helpers ─────────────────────────────────────────────────────────────────

async function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

const TEST_TIMEOUT = 10000; // 10s per test

async function runTest(fn) {
  const timer = setTimeout(() => {
    throw new Error(`${fn.name} timed out after ${TEST_TIMEOUT}ms`);
  }, TEST_TIMEOUT);
  try {
    await fn();
  } finally {
    clearTimeout(timer);
  }
}

// ── Raw TCP pre-check (bypasses client entirely) ────────────────────────────

function rawTcpPing(host, port) {
  return new Promise((resolve, reject) => {
    const socket = net.createConnection({ host, port }, () => {
      // Connected — send RESP PING
      socket.write(Buffer.from('*1\r\n$4\r\nPING\r\n'));
    });

    const timer = setTimeout(() => {
      socket.destroy();
      reject(new Error(`Raw TCP PING timed out after 5000ms`));
    }, 5000);

    let buf = Buffer.alloc(0);
    socket.on('data', (chunk) => {
      buf = Buffer.concat([buf, chunk]);
      // We expect +PONG\r\n (7 bytes) — wait for at least that
      if (buf.length >= 5) {
        clearTimeout(timer);
        socket.destroy();
        resolve(buf.toString('utf8').trim());
      }
    });

    socket.on('error', (err) => {
      clearTimeout(timer);
      reject(err);
    });

    socket.on('close', () => {
      clearTimeout(timer);
      if (buf.length === 0) {
        reject(new Error('Connection closed before any data received'));
      }
    });
  });
}

// ── Core ─────────────────────────────────────────────────────────────────────

async function testPing() {
  const c = await makeClient();
  try {
    const r = await c.ping();
    check('PING', r === 'PONG', `got ${JSON.stringify(r)}`);
  } finally { c.close(); }
}

async function testEcho() {
  const c = await makeClient();
  try {
    const r = await c.echo('hello');
    check('ECHO', r === 'hello', `got ${JSON.stringify(r)}`);
  } finally { c.close(); }
}

async function testDbsize() {
  const c = await makeClient();
  try {
    const n = await c.dbsize();
    check('DBSIZE', typeof n === 'number' && n >= 0, `got ${n}`);
  } finally { c.close(); }
}

// ── String ───────────────────────────────────────────────────────────────────

async function testSetGet() {
  const c = await makeClient();
  try {
    await c.del('n:sg');
    await c.set('n:sg', 'world');
    const v = await c.get('n:sg');
    check('SET/GET', v === 'world', `got ${JSON.stringify(v)}`);
    await c.del('n:sg');
  } finally { c.close(); }
}

async function testGetMissing() {
  const c = await makeClient();
  try {
    const v = await c.get('n:nonexistent_xyz');
    check('GET missing', v === null, `got ${JSON.stringify(v)}`);
  } finally { c.close(); }
}

async function testSetNX() {
  const c = await makeClient();
  try {
    await c.del('n:nx');
    await c.set('n:nx', 'orig');
    const r = await c.set('n:nx', 'other', { nx: true });
    check('SETNX existing', r === null, `got ${JSON.stringify(r)}`);
    await c.del('n:nx');
    const r2 = await c.set('n:nx', 'new', { nx: true });
    check('SETNX missing', r2 === 'OK', `got ${JSON.stringify(r2)}`);
    await c.del('n:nx');
  } finally { c.close(); }
}

async function testSetXX() {
  const c = await makeClient();
  try {
    await c.del('n:xx');
    const r = await c.set('n:xx', 'v', { xx: true });
    check('SETXX missing', r === null, `got ${JSON.stringify(r)}`);
    await c.set('n:xx', 'orig');
    const r2 = await c.set('n:xx', 'upd', { xx: true });
    check('SETXX existing', r2 === 'OK', `got ${JSON.stringify(r2)}`);
    const v = await c.get('n:xx');
    check('SETXX value', v === 'upd', `got ${JSON.stringify(v)}`);
    await c.del('n:xx');
  } finally { c.close(); }
}

async function testDelMulti() {
  const c = await makeClient();
  try {
    await c.set('n:d1', 'a'); await c.set('n:d2', 'b'); await c.set('n:d3', 'c');
    const n = await c.del('n:d1', 'n:d2', 'n:d3');
    check('DEL multi', n === 3, `got ${n}`);
    const v = await c.get('n:d2');
    check('DEL verify', v === null, 'still exists');
  } finally { c.close(); }
}

async function testIncrDecr() {
  const c = await makeClient();
  try {
    await c.del('n:inc');
    check('INCR init', await c.incr('n:inc') === 1, '');
    check('INCRBY', await c.incrBy('n:inc', 9) === 10, '');
    check('DECR', await c.decr('n:inc') === 9, '');
    check('DECRBY', await c.decrBy('n:inc', 4) === 5, '');
    await c.del('n:inc');
  } finally { c.close(); }
}

async function testMSetMGet() {
  const c = await makeClient();
  try {
    await c.mset({ 'n:m1': 'a', 'n:m2': 'b', 'n:m3': 'c' });
    const vals = await c.mget('n:m1', 'n:m2', 'n:m3', 'n:missing');
    check('MSET/MGET', vals[0] === 'a' && vals[1] === 'b' && vals[2] === 'c', `got ${JSON.stringify(vals)}`);
    check('MGET null', vals[3] === null, 'not null');
    await c.del('n:m1', 'n:m2', 'n:m3');
  } finally { c.close(); }
}

async function testAppendStrlen() {
  const c = await makeClient();
  try {
    await c.del('n:ap'); await c.set('n:ap', 'hello');
    const n = await c.append('n:ap', ' world');
    check('APPEND', n === 11, `got ${n}`);
    check('STRLEN', await c.strlen('n:ap') === 11, '');
    await c.del('n:ap');
  } finally { c.close(); }
}

async function testGetRangeSetRange() {
  const c = await makeClient();
  try {
    await c.del('n:gr'); await c.set('n:gr', 'Hello, World!');
    check('GETRANGE', (await c.getRange('n:gr', 0, 4)) === 'Hello', '');
    const n = await c.setRange('n:gr', 7, 'FastKV');
    check('SETRANGE len', n === 13, `got ${n}`);
    check('SETRANGE val', (await c.get('n:gr')) === 'Hello, FastKV', '');
    await c.del('n:gr');
  } finally { c.close(); }
}

async function testExists() {
  const c = await makeClient();
  try {
    await c.del('n:ex1', 'n:ex2'); await c.set('n:ex1', 'v');
    check('EXISTS', await c.exists('n:ex1', 'n:ex2') === 1, '');
    await c.del('n:ex1');
  } finally { c.close(); }
}

// ── TTL ──────────────────────────────────────────────────────────────────────

async function testExpireTtlPersist() {
  const c = await makeClient();
  try {
    await c.del('n:ttl'); await c.set('n:ttl', 'data');
    check('TTL no-expiry', await c.ttl('n:ttl') === -1, '');
    await c.expire('n:ttl', 60);
    const ttl = await c.ttl('n:ttl');
    check('TTL after EXPIRE', ttl > 0 && ttl <= 60, `got ${ttl}`);
    await c.persist('n:ttl');
    check('TTL after PERSIST', await c.ttl('n:ttl') === -1, '');
    await c.del('n:ttl');
  } finally { c.close(); }
}

async function testPttl() {
  const c = await makeClient();
  try {
    await c.del('n:pttl'); await c.set('n:pttl', 'data'); await c.expire('n:pttl', 60);
    const pt = await c.pttl('n:pttl');
    check('PTTL', pt > 0 && pt <= 60000, `got ${pt}`);
    await c.del('n:pttl');
  } finally { c.close(); }
}

// ── Hash ─────────────────────────────────────────────────────────────────────

async function testHashCRUD() {
  const c = await makeClient();
  try {
    await c.del('n:h');
    check('HSET new', await c.hset('n:h', 'name', 'Alice') === 1, '');
    check('HSET update', await c.hset('n:h', 'name', 'Bob') === 0, '');
    check('HGET', (await c.hget('n:h', 'name')) === 'Bob', '');
    check('HEXISTS', await c.hexists('n:h', 'name') === 1, '');
    check('HLEN', await c.hlen('n:h') === 1, '');
    await c.del('n:h');
  } finally { c.close(); }
}

async function testHMSetHGetAll() {
  const c = await makeClient();
  try {
    await c.del('n:hm');
    await c.hmset('n:hm', { f1: 'v1', f2: 'v2', f3: 'v3' });
    const m = await c.hgetAll('n:hm');
    check('HGETALL', Object.keys(m).length === 3 && m.f1 === 'v1', JSON.stringify(m));
    const vals = await c.hmget('n:hm', 'f1', 'f3');
    check('HMGET', vals[0] === 'v1' && vals[1] === 'v3', JSON.stringify(vals));
    await c.del('n:hm');
  } finally { c.close(); }
}

async function testHDelHKeysHVals() {
  const c = await makeClient();
  try {
    await c.del('n:hd');
    await c.hmset('n:hd', { a: '1', b: '2', c: '3' });
    check('HDEL', await c.hdel('n:hd', 'b') === 1, '');
    check('HKEYS', (await c.hkeys('n:hd')).length === 2, '');
    check('HVALS', (await c.hvals('n:hd')).length === 2, '');
    await c.del('n:hd');
  } finally { c.close(); }
}

// ── List ─────────────────────────────────────────────────────────────────────

async function testListPushPop() {
  const c = await makeClient();
  try {
    await c.del('n:l');
    await c.rpush('n:l', 'a', 'b', 'c');
    await c.lpush('n:l', 'z');
    const items = await c.lrange('n:l', 0, -1);
    check('LPUSH+RPUSH', items.length === 4 && items[0] === 'z', JSON.stringify(items));
    check('LPOP', (await c.lpop('n:l')) === 'z', '');
    check('RPOP', (await c.rpop('n:l')) === 'c', '');
    await c.del('n:l');
  } finally { c.close(); }
}

async function testListLTrimLSetLRem() {
  const c = await makeClient();
  try {
    await c.del('n:lt');
    await c.rpush('n:lt', 'a', 'b', 'c', 'd', 'e');
    await c.ltrim('n:lt', 1, 3);
    check('LTRIM', (await c.lrange('n:lt', 0, -1)).length === 3, '');
    await c.lset('n:lt', 0, 'X');
    check('LSET+LINDEX', (await c.lindex('n:lt', 0)) === 'X', '');
    await c.rpush('n:lt', 'b', 'b');
    check('LREM', await c.lrem('n:lt', 2, 'b') === 2, '');
    await c.del('n:lt');
  } finally { c.close(); }
}

async function testListLenIndex() {
  const c = await makeClient();
  try {
    await c.del('n:ll');
    await c.rpush('n:ll', 'x', 'y');
    check('LLEN', await c.llen('n:ll') === 2, '');
    check('LINDEX -1', (await c.lindex('n:ll', -1)) === 'y', '');
    await c.del('n:ll');
  } finally { c.close(); }
}

// ── Pipeline (dedicated connection) ──────────────────────────────────────────

async function testPipeline() {
  const c = await makeClient();
  try {
    await c.del('n:p1', 'n:p2', 'n:p3');
    const pipe = await c.pipeline();
    try {
      pipe.set('n:p1', '10');
      pipe.set('n:p2', '20');
      pipe.incr('n:p1');
      pipe.get('n:p1');
      pipe.get('n:p2');
      pipe.dbsize();
      const results = await pipe.execute();
      check('Pipeline count', results.length === 6, `got ${results.length}`);
      check('Pipeline INCR', results[3] === '11', `got ${JSON.stringify(results[3])}`);
      check('Pipeline GET', results[4] === '20', `got ${JSON.stringify(results[4])}`);
    } finally {
      pipe.close();
    }
    await c.del('n:p1', 'n:p2', 'n:p3');
  } finally { c.close(); }
}

// ── WRONGTYPE ────────────────────────────────────────────────────────────────

async function testWrongType() {
  const c = await makeClient();
  try {
    await c.del('n:wt');
    await c.set('n:wt', 'string');
    try {
      await c.llen('n:wt');
      check('WRONGTYPE', false, 'should have thrown');
    } catch (e) {
      check('WRONGTYPE', e.message && e.message.includes('WRONGTYPE'), e.message);
    }
    await c.del('n:wt');
  } finally { c.close(); }
}

// ── Concurrent ───────────────────────────────────────────────────────────────

async function testConcurrent() {
  const c = await makeClient();
  try {
    const promises = [];
    for (let i = 0; i < 10; i++) {
      promises.push((async (tid) => {
        await c.set(`n:conc:${tid}`, 'v');
        await c.get(`n:conc:${tid}`);
      })(i));
    }
    await Promise.all(promises);
    check('Concurrent', true, '');
    await c.del(...Array.from({ length: 10 }, (_, i) => `n:conc:${i}`));
  } finally { c.close(); }
}

// ── Main ─────────────────────────────────────────────────────────────────────

async function main() {
  console.log(`FastKV Node.js Integration Tests — ${HOST}:${PORT}`);
  console.log('='.repeat(55));

  // Pre-check: raw TCP PING (bypasses client code entirely)
  console.log('  [diag] Raw TCP PING ...');
  try {
    const raw = await rawTcpPing(HOST, PORT);
    console.log(`  [diag] Raw TCP OK — server replied: ${JSON.stringify(raw)}`);
  } catch (e) {
    console.error(`  [diag] Raw TCP FAILED — ${e.message}`);
    console.error('  [diag] Skipping all client tests (server unreachable from Node.js net)');
    process.exit(1);
  }

  const tests = [
    testPing, testEcho, testDbsize,
    testSetGet, testGetMissing, testSetNX, testSetXX,
    testDelMulti, testIncrDecr, testMSetMGet,
    testAppendStrlen, testGetRangeSetRange, testExists,
    testExpireTtlPersist, testPttl,
    testHashCRUD, testHMSetHGetAll, testHDelHKeysHVals,
    testListPushPop, testListLTrimLSetLRem, testListLenIndex,
    testPipeline, testWrongType, testConcurrent,
  ];

  for (const fn of tests) {
    try {
      await runTest(fn);
    } catch (e) {
      failed++;
      console.error(`  FAIL  ${fn.name} — ${e.message}`);
    }
  }

  console.log('='.repeat(55));
  console.log(`Results: ${passed} passed, ${failed} failed, ${passed + failed} total`);
  process.exit(failed > 0 ? 1 : 0);
}

main().catch(e => { console.error(e); process.exit(1); });

File: clients/python/tests/test_integration.py
#!/usr/bin/env python3
"""
FastKV Python client — integration tests.

Runs against a live FastKV server.  Set FASTKV_HOST / FASTKV_PORT
to override the defaults (localhost:8379).
"""

from __future__ import annotations

import os
import sys
import threading
import traceback

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from fastkv.client import FastKVClient
from fastkv.exceptions import FastKVResponseError, FastKVConnectionError

HOST = os.environ.get("FASTKV_HOST", "localhost")
PORT = int(os.environ.get("FASTKV_PORT", "6379"))

passed = 0
failed = 0
errors: list[str] = []


def run(name: str, fn) -> None:
    global passed, failed
    try:
        fn()
        passed += 1
        print(f"  PASS  {name}")
    except Exception:
        failed += 1
        tb = traceback.format_exc()
        errors.append(f"FAIL {name}:\n{tb}")
        print(f"  FAIL  {name}")


# ── helpers ──────────────────────────────────────────────────────────────────

def client() -> FastKVClient:
    return FastKVClient(HOST, PORT, socket_timeout=5)


# ── Core ─────────────────────────────────────────────────────────────────────

def test_ping():
    with client() as c:
        r = c.ping()
        assert r == b"PONG" or r == "PONG", f"got {r!r}"

def test_echo():
    with client() as c:
        r = c.echo("hello")
        assert r == b"hello" or r == "hello", f"got {r!r}"

def test_dbsize():
    with client() as c:
        n = c.dbsize()
        assert isinstance(n, int)

# ── String ───────────────────────────────────────────────────────────────────

def test_set_get():
    with client() as c:
        c.set("py:sg", "world")
        assert c.get("py:sg") == b"world"
        c.delete("py:sg")

def test_get_missing():
    with client() as c:
        assert c.get("py:nonexistent_xyz") is None

def test_setnx():
    with client() as c:
        c.delete(["py:nx"])
        assert c.set("py:nx", "orig") is True
        r = c.set("py:nx", "new", nx=True)
        assert r is None  # NX failed — key exists
        assert c.get("py:nx") == b"orig"
        c.delete(["py:nx"])
        r = c.set("py:nx", "new", nx=True)
        assert r is True
        c.delete(["py:nx"])

def test_setxx():
    with client() as c:
        c.delete(["py:xx"])
        r = c.set("py:xx", "v", xx=True)
        assert r is None  # XX on missing key
        c.set("py:xx", "orig")
        r = c.set("py:xx", "upd", xx=True)
        assert r is True
        assert c.get("py:xx") == b"upd"
        c.delete(["py:xx"])

def test_del_multi():
    with client() as c:
        for k in ("py:d1", "py:d2", "py:d3"):
            c.set(k, "v")
        n = c.delete(["py:d1", "py:d2", "py:d3"])
        assert n == 3, f"DEL returned {n}"
        assert c.get("py:d2") is None

def test_del_single():
    with client() as c:
        c.set("py:ds", "v")
        n = c.delete("py:ds")
        assert n == 1
        assert c.get("py:ds") is None

def test_incr_decr():
    with client() as c:
        c.delete(["py:inc"])
        assert c.incr("py:inc") == 1  # init
        assert c.incrby("py:inc", 9) == 10
        assert c.decr("py:inc") == 9
        assert c.decrby("py:inc", 4) == 5
        c.delete(["py:inc"])

def test_mset_mget():
    with client() as c:
        c.mset({"py:m1": "a", "py:m2": "b", "py:m3": "c"})
        vals = c.mget(["py:m1", "py:m2", "py:m3"])
        assert vals == [b"a", b"b", b"c"], f"got {vals!r}"
        c.delete(["py:m1", "py:m2", "py:m3"])

def test_append_strlen():
    with client() as c:
        c.delete(["py:ap"])
        c.set("py:ap", "hello")
        assert c.append("py:ap", " world") == 11
        assert c.strlen("py:ap") == 11
        c.delete(["py:ap"])

def test_getrange():
    with client() as c:
        c.delete(["py:gr"])
        c.set("py:gr", "Hello, World!")
        r = c.getrange("py:gr", 0, 4)
        assert r == b"Hello", f"got {r!r}"
        c.delete(["py:gr"])

def test_setrange():
    with client() as c:
        c.delete(["py:sr"])
        c.set("py:sr", "Hello, World!")
        r = c.setrange("py:sr", 7, "FastKV")
        assert r == 13, f"SETRANGE returned {r}"
        v = c.get("py:sr")
        assert v == b"Hello, FastKV", f"got {v!r}"
        c.delete(["py:sr"])

def test_exists():
    with client() as c:
        c.delete(["py:ex1", "py:ex2"])
        c.set("py:ex1", "v")
        n = c.exists(["py:ex1", "py:ex2"])
        assert n == 1, f"EXISTS returned {n}"
        c.delete(["py:ex1"])

# ── TTL ──────────────────────────────────────────────────────────────────────

def test_expire_ttl_persist():
    with client() as c:
        c.delete(["py:ttl"])
        c.set("py:ttl", "data")
        assert c.ttl("py:ttl") == -1
        assert c.expire("py:ttl", 60) is True
        ttl = c.ttl("py:ttl")
        assert 1 <= ttl <= 60, f"TTL={ttl}"
        assert c.persist("py:ttl") is True
        assert c.ttl("py:ttl") == -1
        c.delete(["py:ttl"])

def test_pttl():
    with client() as c:
        c.delete(["py:pttl"])
        c.set("py:pttl", "data")
        c.expire("py:pttl", 60)
        pt = c.pttl("py:pttl")
        assert 1000 <= pt <= 60000, f"PTTL={pt}"
        c.delete(["py:pttl"])

# ── Hash ─────────────────────────────────────────────────────────────────────

def test_hash_crud():
    with client() as c:
        c.delete(["py:h"])
        assert c.hset("py:h", "name", "Alice") == 1
        assert c.hset("py:h", "name", "Bob") == 0  # update
        assert c.hget("py:h", "name") == b"Bob"
        assert c.hexists("py:h", "name") is True
        assert c.hlen("py:h") == 1
        c.delete(["py:h"])

def test_hmset_hgetall():
    with client() as c:
        c.delete(["py:hm"])
        c.hmset("py:hm", {"f1": "v1", "f2": "v2", "f3": "v3"})
        m = c.hgetall("py:hm")
        assert len(m) == 3, f"HGETALL got {len(m)} fields"
        assert m[b"f1"] == b"v1"
        vals = c.hmget("py:hm", "f1", "f3")
        assert vals == [b"v1", b"v3"], f"got {vals!r}"
        c.delete(["py:hm"])

def test_hdel_hkeys_hvals():
    with client() as c:
        c.delete(["py:hd"])
        c.hmset("py:hd", {"a": "1", "b": "2", "c": "3"})
        assert c.hdel("py:hd", "b") == 1
        assert len(c.hkeys("py:hd")) == 2
        assert len(c.hvals("py:hd")) == 2
        c.delete(["py:hd"])

# ── List ─────────────────────────────────────────────────────────────────────

def test_list_push_pop():
    with client() as c:
        c.delete(["py:l"])
        c.rpush("py:l", "a", "b", "c")
        c.lpush("py:l", "z")
        items = c.lrange("py:l", 0, -1)
        assert items == [b"z", b"a", b"b", b"c"], f"got {items}"
        assert c.lpop("py:l") == b"z"
        assert c.rpop("py:l") == b"c"
        c.delete(["py:l"])

def test_list_ltrim_lset_lrem():
    with client() as c:
        c.delete(["py:lt"])
        c.rpush("py:lt", "a", "b", "c", "d", "e")
        assert c.ltrim("py:lt", 1, 3) is True
        items = c.lrange("py:lt", 0, -1)
        assert len(items) == 3
        assert c.lset("py:lt", 0, "X") is True
        assert c.lindex("py:lt", 0) == b"X"
        c.rpush("py:lt", "b", "b")
        assert c.lrem("py:lt", 2, "b") == 2
        c.delete(["py:lt"])

def test_list_len_index():
    with client() as c:
        c.delete(["py:ll"])
        c.rpush("py:ll", "x", "y")
        assert c.llen("py:ll") == 2
        assert c.lindex("py:ll", -1) == b"y"
        c.delete(["py:ll"])

# ── Pipeline ─────────────────────────────────────────────────────────────────

def test_pipeline():
    with client() as c:
        c.delete(["py:p1", "py:p2", "py:p3"])
        with c.pipeline() as p:
            p.set("py:p1", "10")
            p.set("py:p2", "20")
            p.incr("py:p1")
            p.get("py:p1")
            p.get("py:p2")
            p.mget(["py:p1", "py:p2", "py:p3"])
            results = p.execute()

        assert len(results) == 6, f"expected 6 results, got {len(results)}"
        assert results[3] == b"11", f"INCR result: {results[3]!r}"
        assert results[4] == b"20", f"GET result: {results[4]!r}"
        c.delete(["py:p1", "py:p2", "py:p3"])

def test_pipeline_backpressure():
    with client() as c:
        from fastkv.pipeline import Pipeline
        p = Pipeline(c, max_size=100)
        for i in range(200):
            p.get(f"py:bp:{i}")
        try:
            p.execute()
            assert False, "should have raised ValueError"
        except ValueError as e:
            assert "exceeds" in str(e).lower()

# ── WRONGTYPE ────────────────────────────────────────────────────────────────

def test_wrongtype():
    with client() as c:
        c.delete(["py:wt"])
        c.set("py:wt", "string")
        try:
            c.llen("py:wt")
            assert False, "should raise"
        except FastKVResponseError as e:
            assert "WRONGTYPE" in str(e)
        c.delete(["py:wt"])

# ── Concurrent access ────────────────────────────────────────────────────────

def test_concurrent():
    errs = []

    def worker(tid):
        try:
            with client() as c:
                c.set(f"py:conc:{tid}", "v")
                c.get(f"py:conc:{tid}")
        except Exception as e:
            errs.append(e)

    threads = [threading.Thread(target=worker, args=(i,)) for i in range(10)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    assert not errs, f"concurrent errors: {errs}"
    with client() as c:
        c.delete([f"py:conc:{i}" for i in range(10)])


# ── Main ─────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    print(f"FastKV Python Integration Tests — {HOST}:{PORT}")
    print("=" * 55)

    tests = [v for k, v in sorted(globals().items()) if k.startswith("test_")]
    for fn in tests:
        run(fn.__name__, fn)

    print("=" * 55)
    print(f"Results: {passed} passed, {failed} failed, {passed + failed} total")
    if errors:
        print("\nFailures:")
        for e in errors:
            print(e)
    sys.exit(1 if failed else 0)
